AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -796,17 +798,142 @@ - (BOOL)shouldCalculateLayoutWithConstrainedSize:(ASSizeRange)constrainedSize pa
   return _calculatedDisplayNodeLayout->isValidForConstrainedSizeParentSize(constrainedSize, parentSize) == NO;
 }
 
-- (ASLayoutElementType)layoutElementType
+// TODO: coalayout: Deprecate API instead call -[ASDisplayNode setNeedsLayout]
+- (void)invalidateCalculatedLayout
 {
-  return ASLayoutElementTypeDisplayNode;
+  [self __setNeedsLayout];
 }
 
-- (BOOL)canLayoutAsynchronous
+/**
+ * Invalidate the current calculated layout so it can be recalculated in the next layout pass
+ */
+- (void)__setNeedsLayout
 {
-  return !self.isNodeLoaded;
+  ASDN::MutexLocker l(__instanceLock__);
+
+  _calculatedDisplayNodeLayout->invalidate();
+  _pendingDisplayNodeLayout->invalidate();
 }
 
-#pragma mark - Automatic Hierarchy
+/**
+ * Returns a BOOL indicating whether the layer has been marked as needing a layout update.
+ */
+- (BOOL)__needsLayout
+{
+  return _calculatedDisplayNodeLayout->isDirty();
+}
+
+/**
+ * The node's supernodes are traversed until a ancestor node is found that does not require layout. Then layout
+ * is performed on the entire node-tree beneath that ancestor
+ */
+- (void)__layoutIfNeeded
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+  
+  // Invalidate the current calculated layout so in the next layout pass it get's recalculated
+  [self invalidateCalculatedLayout];
+  
+  // Walk the node's supernodes until a ancestor node is found that does not require layout
+  if ([_supernode __needsLayout]) {
+    ASDisplayNode *supernode = _supernode;
+    __instanceLock__.unlock();
+    [_supernode __layoutIfNeeded];
+    return;
+  }
+
+  // Layout all subviews starting from the first node that needs layout
+  [self __layout];
+  
+  __instanceLock__.unlock();
+}
+
+/**
+ * Called from [CALayer layoutSublayers:]: All subnode layout related code should happen in there
+ */
+- (void)__layout
+{
+  ASDisplayNodeAssertMainThread();
+  ASDN::MutexLocker l(__instanceLock__);
+  CGRect bounds = _threadSafeBounds;
+  
+  // Prevent zero bounds subnode adjustments
+  if (CGRectEqualToRect(bounds, CGRectZero)) {
+    // Performing layout on a zero-bounds view often results in frame calculations with negative sizes after applying
+    // margins, which will cause layoutThatFits: on subnodes to assert.
+    LOG(@"Warning: No size given for node before node was trying to layout itself: %@. Please provide a frame for the node.", self);
+    return;
```

Comment:
The `bounds` change of a note is actually an interesting problem ... if you change the frame of a `UIView` it also invalidates the layout of the superview. We should have the same behavior in this case. The question is just how do we prevent any infinite loops as we set the frames of the subnodes within `__layoutSublayouts` that would invalidate the supernode. I guess we only invalidate the layout if the `bounds` actually changes.